### PR TITLE
#160314849 Users should be able to categorise their articles

### PR DESCRIPTION
**Pull request by emmaadesile**
### What does this PR do?
Users can categorize their articles

### Description of tasks to be completed
- Admins can do CRUD operations on categories
- Users can get all categories
- Users can view all the articles in a category
- Authors can add their articles to a category
- Authors can remove their articles from any category

### How to manually test it
- After cloning the repo
- Install and configure Postgresql
- use .envsample to create your environment variables
- run `npm install` to install dependencies
- run `npm run migrate` to create the create tables in the db
- run `npm start:dev` to start the development server
- Admin can create `POST` a new `category/api/article-categories`
- Admin can update `PUT` an existing category `/api/article-categories/:categoryName`
- Admin can delete `DELETE` an existing category `/api/article-categories/:categoryName`
- A user or an author can view all categories `GET` `/api/article-categories`
- A user or an author can view all articles in a category GET /api/article-categories
- Authors can add their article(s) to a category using the articleId `POST` `/api/article-categories/:categoryName`

### What are the relevant pivotal tracker stories?
[160314849](https://www.pivotaltracker.com/story/show/160314849)

### CODE REVIEW

**Review by d-beloved — APPROVED** [review]
LGTM

**Comment by CEOehis** [comment]
LGTM

**Review by CEOehis — APPROVED** [review]
LGTM

File: server/controllers/categoryController.js [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+  }
+
+  /**
+   * @description Create a new category
+   * @param  {object} req body of the Admin's request
+   * @param  {function} res response from the server
+   * @returns {object} The body of the response message
+   * @memberof CategoryController
+   */
+  static createCategory(req, res) {
```
Comment (aknwosu) [inline server/controllers/categoryController.js:39]: this should only be accessible by admins

File: server/controllers/categoryController.js [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+    const newCategory = req.body.name.trim();
+    if (newCategory === '') {
+      return res.status(400).json({
+        status: 'error',
+        error: 'Name field cannot be empty'
+      });
+    }
+    return Category.all()
+      .then((allCategories) => {
+        if (allCategories.length === 10) {
```
Comment (aknwosu) [inline server/controllers/categoryController.js]: why is there a limit on how many categories can be added?

File: server/controllers/categoryController.js [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+          return res.status(400).json({
+            status: 'error',
+            error: 'You cannot have more that 10 article categories'
+          });
+        }
+        return Category.findOne({
+          where: { name: newCategory }
```
Comment (aknwosu) [inline server/controllers/categoryController.js]: you probably want to sanitize this and make sure categories cannot be duplicated depending on the capitalization

File: server/controllers/categoryController.js [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+      .then((category) => {
+        if (!category) {
+          return res.status(404).json({
+            status: 'error',
+            error: 'Category does not exist'
+          });
+        }
+        // If category exists, check if article exists
+        Article.findOne(({
+          where: { title: articleTitle }
```
Comment (aknwosu) [inline server/controllers/categoryController.js]: use article ID

File: server/controllers/categoryController.js [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+        model: Article,
+        as: 'category',
+        attributes: { exclude: ['createdAt', 'updatedAt'] }
+      }],
+    })
+      .then((category) => {
+        if (!category) {
+          return res.status(404).json({
+            status: 'error',
+            error: 'Category does not exist'
```
Comment (aknwosu) [inline server/controllers/categoryController.js]: error: 'There are no articles in this category'

File: server/controllers/categoryController.js [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+          });
+        }
+        // If category exists, check if article exists
+        Article.findOne(({
+          where: { title: articleTitle }
```
Comment (aknwosu) [inline server/controllers/categoryController.js]: use articleID

File: server/controllers/categoryController.js [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+            const categoryId = category.id;
+            const articleId = article.id;
+            const { authorization } = req.headers;
+            const token = authorization.split(' ')[1];
+            const decoded = TokenHelper.decodeToken(token);
+            const userId = decoded.id;
+            if (article.authorId !== userId) {
+              return res.status(403).json({
+                status: 'error',
+                error: 'You cannot remove another author\'s article from this category',
```
Comment (aknwosu) [inline server/controllers/categoryController.js]: "You do not have sufficient permissions for this action"

**Pull request merged**